[WIP] Dmitry Grapov | Iterations 1-5

Dmitry-Grapov opened the pull request:
Python Final Task.

- [x] First iteration
  - [x] -h, --help show this help message and exit
  - [x] --version Print version info
  - [x] --json Print result as JSON in stdout
  - [x] --verbose Outputs verbose status messages
  - [x] --limit LIMIT Limit news topics if this parameter provided
 - [x] Second iteration
 - [x] Third iteration
 - [x] Fourth iteration
 - [x] Fifth iteration
 - [ ] Sixth iteration

Dmitry-Grapov committed:
CLI optional and positional arguments added
Parsing of CLI arguments provided
Limited RSS parsing added
setuptools setup.py configured

Dmitry-Grapov committed:
1. Logging added
2. requirements.txt up to date
3. Error message for the faulty URLs added

Dmitry-Grapov committed:
1. Limit range check added in main.py
2. RSS printing function  added to rss_parser.py

Dmitry-Grapov committed:
1. Module disp.py removed as unnecessary
2. Local time conversion implemented
3. rss_parser.py module cleared of unnecessary functions

Dmitry-Grapov committed:
1. Docstring added to the rss_parser.py module
2. Printed links are marked as (link) or (image)
3. Added function html2text which converts html document to plain text. The funciton is based on BeautifulSoup4.
4. BeautifulSoup is used to find links in the Summary and Content of RSS

Dmitry-Grapov committed:
1. JSON conversion and print added
2. Link and image html tags substituted by referencies

Dmitry-Grapov committed:
1. rss_parser.py RssParser class use simplified. Create instance -> check wherther is_empty() -> if not empty: dump_json() or print_json()
2. find_links() method removed from class RssParser as redundant
3. main.py main() method was adopted for simplified RssParser

Dmitry-Grapov committed:
1. Docstring and comments improved
2. setup.py - program version updated
3. When printing JSON data, cyrillic is displayed as letters

Dmitry-Grapov committed:
1. Docstrings updated
2. Time in JSON now is ISO sting
3. Redundant JSON entries deleted: short_date, short_time, url

Dmitry-Grapov committed:
1. Removed 'Content' key from JSON
2. Improved error handling for empty dates from RSS feed
3. Docstring updated and extended

Dmitry-Grapov committed:
1. requrerements.txt updated.
2. rss_reader.py module split on 4 modules: xml_downloader.py xml_to_json.py json_to_json.py json_io.py
3. Unit test for the new modules added: test_xml_downloader.py test_xml_to_json.py test_json_to_json.py test_json_io.py
4. In main() module added optional argument args.date.
5. In main() module args.source argument mandatory now.

Dmitry-Grapov committed:
1. xml_to_json module in html_json_entry dictionary new key added: "URL".
2. json_io module: JSON entries whithout date are now skipped instead of being saved.
3. json_io module: new method find_raw_rss() to seek for JSON files from a specific date.
4. json_io module docstrings updated

Dmitry-Grapov committed:
1. test_xml_to_json and test_json_io updated
2. requirements.txt and setup.py edited - removed dependency pkg_resourses=0.0.0
3. json_io module modified to load data whith empty URL argument
3. xml_to_json module updated to work with new dictionary key - "URL"

Dmitry-Grapov committed:
Iteration 3

Dmitry-Grapov committed:
1. In the module xml_to_json.py timezone conversion code fixed.
2. Module test_xml_to_json.py modified to work in different timezones.

Dmitry-Grapov committed:
1. Modified installation script .travis.yml: installation of pycodestyle and termcolor added.

Dmitry-Grapov committed:
1. In travis.yml installation of the rss_reader module added.

Dmitry-Grapov committed:
Merge branch 'main' into main

Dmitry-Grapov committed:
Long strings and dictionaries are reformatted to match PEP8 style.

Dmitry-Grapov committed:
Merge branch 'main' of https://github.com/Dmitry-Grapov/PythonFinalTask

Shurke commented:
К сожалению, не работает rss_reader.
![Screenshot 2021-06-12 at 14 23 30](https://user-images.githubusercontent.com/44495358/121774186-ca4e0180-cb89-11eb-8c7a-d2637ec5cb54.png)

Shurke commented on .travis.yml:
```diff
@@ -3,10 +3,17 @@ python:
   - python ./../rss_reader_checker.py
-  - cd -
\ No newline at end of file
+<<<<<<< HEAD
+  - cd -
+=======
+  - cd -
+
+>>>>>>> 643f03232677a04e50332d838b1744c6efa37982
```
Вы закомитили файл, в котором есть конфликт.
Исправьте, чтобы CI работал корректно.

Dmitry-Grapov committed:
An attempt to resolve conflist in travis.yaml

Dmitry-Grapov committed:
1. Module xml_to_json modified to parse date in different formats. Dateparser library used.
2. In module test_xml_to_json test data modified: EST time was mistakenly considered -0400, now it is -0500 timezone.
3. In requirement.txt and setup.py new dependency added: dateparser==1.0.0

Dmitry-Grapov committed:
1. Module xml_to_json modified to parse date in different formats. Dateparser library used.
2. In module test_xml_to_json test data modified: EST time was mistakenly considered -0400, now it is -0500 timezone.
3. In requirement.txt and setup.py new dependency added: dateparser==1.0.0

Dmitry-Grapov committed:
Merge branch 'main' of https://github.com/Dmitry-Grapov/PythonFinalTask

Dmitry-Grapov committed:
1. Program version updated to 1.4 in setup.py main.py and xml_downloader.py.
2. In travis.yml added installation of wkhtmltopdf utility.
3. In main.py new options added --to-html --to-pdf.
4. New modules added: json_to_html and test_json_to_html. Modules used for JSON conversion to HTML fromat and file saving.

Dmitry-Grapov committed:
1. Modified installation of wkhtmltopdf utility.

Dmitry-Grapov committed:
1.In travis.yml installing wkhtmltopdf with apt-get.

Dmitry-Grapov committed:
1. In travis.yml added installation of wkhtmltopdf utility with apt-get.

Dmitry-Grapov committed:
Commit for rss_reader version 1.5:
1. Added new dependency colored==1.4.2 in requirements.txt and setup.py
2. Version number updated to 1.5 in main.py xml_downloader.py and setup.py
3. In main.py added new argument --colorize and checks for its use.
4. Code for output colorization added to json_to_json.py

Dmitry-Grapov committed:
1. In modules main.py and json_to_html improved error handling in case of wrong save file path provided for html and pdf conversion.
2. In module main.py enabled access to local files for pdfkit library.
3. In module json_to_html path for saved images on Windows platform now starst with file:\\\. It enables pdfkit library properly convert html documents with images saved in local storage.

Shurke commented on rss_reader/json_io.py:
```diff
@@ -0,0 +1,278 @@
+"""json_io module
+
+Classes:
+-------
+    JsonIO
+
+Attributes:
+----------
+    _url
+    _limit
+    _base_directory_path
+
+Methods:
+-------
+    __init__(self, dirname=".rss_reader")
+    get_base_directory_path(self)
+    save_raw_rss(self, raw_json_list)
+    find_raw_rss(self, date: str, url: str = None, limit: int = None)
+    load_raw_rss(self, date: str, url: str)
+    (self, raw_json_list: list, url_dir)
+"""
```
PEP257 описывает правила оформления строк документации.

Shurke commented on rss_reader/json_io.py:
```diff
@@ -0,0 +1,278 @@
+            self._base_directory_path - Directory name for the JSON storage.
+        """
+
+        #  self._limit = limit
```
Если закомментировано, значит не нужно.

Shurke commented on rss_reader/json_io.py:
```diff
@@ -0,0 +1,278 @@
+        self._base_directory_path = os.path.join(home_path, dirname)  # /home/username/.rss_reader
+        logging.info(f"JsonIO:Base directory path:{self._base_directory_path}")
+
+    def get_base_directory_path(self) -> str:
+        """get_base_directory_path() - Getter for self._base_directory_path"""
+        return self._base_directory_path
```
Почему доступ к атрибуту организован через метод?
Если можно запросить сразу атрибут.

Shurke commented on rss_reader/json_io.py:
```diff
@@ -0,0 +1,278 @@
+
+        # When only date provided, seeking JSON entries for this date in all available directories.
+        url_dir = []  # Making list of all available URL directories.
+        for d in os.listdir(self._base_directory_path):
```
d - пример отвратительного нейменга.

Shurke commented on rss_reader/json_io.py:
```diff
@@ -0,0 +1,278 @@
+                url_dir.append(d)
+
+        big_raw_json_list = []
+        for U in url_dir:  # Collecting JSON entries from all available URL directories.
```
U - пример отвратительного нейменга.

Shurke commented on rss_reader/main.py:
```diff
@@ -0,0 +1,153 @@
+        1 - Error
+    """
+    # Processing command line arguments using module argparse
+    parser = argparse.ArgumentParser(description="Pure Python command-line RSS reader.")
+
+    # Adding command line arguments to the application
+    # Optional arguments:
+    parser.add_argument("--version", action="store_true", help="Print version info")
+    parser.add_argument("--json", action="store_true", help="Print result as JSON in stdout")
+    parser.add_argument("--verbose", action="store_true", help="Output verbose status messages")
+    parser.add_argument("--limit", type=int, help="Limit news topics if this parameter provided")
+    parser.add_argument("--date", type=str, help="Print news from specified date YYYYMMDD")
+    parser.add_argument("--to-html", type=str, help="Save news as html")
+    parser.add_argument("--to-pdf", type=str, help="Save news as pdf")
+    parser.add_argument("--colorize", action="store_true", help="Print news in colorized mode")
+    # Positional (mandatory) arguments:
+    parser.add_argument("source", type=str, nargs="?",  help="RSS URL")
+    # Parsing arguments
+    args = parser.parse_args()
+
+    # [--verbose] mode means that status messages are printed while programming is running
+    # logging module is used to print status messages
+    if args.verbose:  # Configuring logging to enable status messages
+        logging.basicConfig(level=logging.INFO, format=" %(asctime)s - %(levelname)s - %(message)s")
+        logging.info("Verbose mode ON")
+    else:  # Configuring logging to disable status messages
+        logging.basicConfig(level=logging.CRITICAL, format=" %(asctime)s - %(levelname)s - %(message)s")
+        logging.disable(logging.CRITICAL)
```
Инициализацию `ArgumentParser` можно было вынести в отдельную функцию.

Shurke commented on rss_reader/main.py:
```diff
@@ -0,0 +1,153 @@
+        exit(1)
+
+    # if args.date exists but includes not only digits or has improper length:
+    if args.date and (not args.date.isdigit() or not len(args.date) == 8):
+        print("DATE must be 8 digits", flush=True)
+        exit(1)
```
В данном случае, дата вида 17171717 будет вполне валидной. 
Но вот только толку от этого нет.

Shurke commented on rss_reader/json_io.py:
```diff
@@ -0,0 +1,278 @@
+
+        logging.info(f"download_images:Reference: {all_links_dict}")
+
+        for key in all_links_dict.keys():
```
Как вариант
```python
for key in all_links_dict:
```
Вот только key - не самая удачная переменная.

Shurke commented on rss_reader/main.py:
```diff
@@ -0,0 +1,153 @@
+        text_json_list.dump_json()
+    else:  # by default: news are printed to stdout as formatted text
+        logging.info("Print RSS in plain text")
+        text_json_list.print_json(color=args.colorize)
```
Странное название метода, для вывода новостей в stdout не в json формате.

Shurke reviewed (commented):
Организация хранения файлов необычная, но на мой взгляд, избыточна сложна.
Нейминг в целом нормальный, но порой возникают вопросы.
